src/remoted/manager.c

```diff
@@ -1070,6 +975,194 @@ STATIC void process_deleted_multi_groups() {
+            }
+        } else {
+            closedir(dir);
+            validate_shared_files(file, group, merged_tmp, _f_sum, f_sum, f_size, create_merged, path_offset);
+        }
+    }
+    free_strarray(files);
+    return;
+}
+
+STATIC void copy_directory(const char *src_path, const char *dst_path, const char *group) {
+    unsigned int i;
+    time_t *modify_time = NULL;
+    int ignored;
+    DIR *dir;
+    char ** files;
+
+    if (files = wreaddir(src_path), !files) {
+        if (errno != ENOTDIR) {
+            mwarn("Could not open directory '%s'. Group folder was deleted.", src_path);
+            return;
+        }
+    }
+
+    for (i = 0; files[i]; ++i) {
+        /* Ignore hidden files  */
+        /* Leave the shared config file for later */
+        /* Also discard merged.mg.tmp */
+        if (files[i][0] == '.' || !strncmp(files[i], SHAREDCFG_FILENAME, strlen(SHAREDCFG_FILENAME))) {
+            continue;
+        }
+
+        char source_path[PATH_MAX + 1] = {0};
+        snprintf(source_path, PATH_MAX + 1, "%s/%s", src_path, files[i]);
+
+        /* Is a file */
+        if (dir = opendir(source_path), !dir) {
+            ignored = 0;
+
+            char destination_path[PATH_MAX + 1] = {0};
+            char agent_conf_chunck_message[PATH_MAX + 1]= {0};
+
+            snprintf(destination_path, PATH_MAX + 1, "%s/%s", dst_path, files[i]);
+            if (modify_time = (time_t*) OSHash_Get(invalid_files, source_path), modify_time != NULL) {
+                ignored = 1;
+            }
+            if (!ignored) {
+                /* If the file is agent.conf, append */
+                if (strcmp(files[i],"agent.conf") == 0) {
+                    snprintf(agent_conf_chunck_message, PATH_MAX + 1,"<!-- Source file: %s/agent.conf -->\n", group);
+                    w_copy_file(source_path, destination_path,'a', agent_conf_chunck_message, 1);
+                } else {
+                    w_copy_file(source_path, destination_path,'c', NULL, 1);
+                }
+            }
+        } else {
+            /* Is a directory */
+            char new_src_path[PATH_MAX + 1] = {0};
+            char new_dst_path[PATH_MAX + 1] = {0};
+            mdebug2("Making new directory: %s", files[i]);
+
+            snprintf(new_src_path, PATH_MAX + 1, "%s/%s", src_path, files[i]);
+            snprintf(new_dst_path, PATH_MAX + 1, "%s/%s", dst_path, files[i]);
+
+            int oldmask = umask(0006);
+            int retval = mkdir(new_dst_path, 0770);
+            umask(oldmask);
+
+            if (retval < 0) {
+                if (errno != EEXIST) {
+                    merror("Cannot create directory '%s': %s (%d)", new_dst_path, strerror(errno), errno);
+                    closedir(dir);
+                    continue;
+                }
+            }
+
+            copy_directory(new_src_path, new_dst_path, group);
+            closedir(dir);
+        }
+    }
+    free_strarray(files);
```

Comment:
Not necessary, but add a return like you did in the previous function to indicate the end of the function.